src/sync.cpp

```diff
@@ -511,12 +511,149 @@ bool assignFilesystemIds(Sync& sync, MegaApp& app, FileSystemAccess& fsaccess, h
     return success;
 }
 
+SyncConfigBag::SyncConfigBag(DbAccess& dbaccess, FileSystemAccess& fsaccess, PrnGen& rng, const std::string& id)
+{
+    std::string dbname = "syncconfigs" + id;
+    mTable.reset(dbaccess.open(rng, &fsaccess, &dbname, false, false));
+    if (!mTable)
+    {
+        LOG_err << "Unable to open DB table: " << dbname;
+        assert(false);
+        return;
+    }
+
+    mTable->rewind();
+
+    uint32_t tableId;
+    std::string data;
+    while (mTable->next(&tableId, &data))
+    {
+        auto syncConfig = SyncConfig::unserialize(data);
+        if (!syncConfig)
+        {
+            LOG_err << "Unable to unserialize sync config at id: " << tableId;
+            assert(false);
+            continue;
+        }
+        syncConfig->dbid = tableId;
+
+        mSyncConfigs.insert(std::make_pair(syncConfig->getLocalPath(), *syncConfig));
+        if (tableId > mTable->nextid)
+        {
+            mTable->nextid = tableId;
+        }
+    }
+    ++mTable->nextid;
+}
+
+void SyncConfigBag::insert(const SyncConfig& syncConfig)
+{
+    auto syncConfigPair = mSyncConfigs.find(syncConfig.getLocalPath());
+    if (syncConfigPair == mSyncConfigs.end())
+    {
+        if (mTable)
+        {
+            std::string data;
+            const_cast<SyncConfig&>(syncConfig).serialize(&data);
+            DBTableTransactionCommitter committer{mTable.get()};
+            if (!mTable->put(mTable->nextid, &data))
+            {
+                LOG_err << "Incomplete database put at id: " << mTable->nextid;
+                assert(false);
+                mTable->abort();
+                return;
+            }
+        }
+        auto insertPair = mSyncConfigs.insert(std::make_pair(syncConfig.getLocalPath(), syncConfig));
+        if (mTable)
+        {
+            insertPair.first->second.dbid = mTable->nextid;
+            ++mTable->nextid;
+        }
+    }
+    else
+    {
+        const auto tableId = syncConfigPair->second.dbid;
+        if (mTable)
+        {
+            DBTableTransactionCommitter committer{mTable.get()};
+            if (!mTable->del(tableId))
```

Comment:
I think you don't need to delete+put here. The `put()` already allows to `insert or update` SQL query :)

My suggestion was to somehow prevent additions that actually update an existing config. The caller should know whether the config is a new one (add) or an update of an existing one (update). Therefore, I was thinking about a `SyncConfigBag::add()` that fails if it already exists, plus a `SyncConfigBag::update()` that fails if the config doesn't exist yet. If you feel it's counterproductive, feel free to go ahead with the new `insert()`.

Reply:
Indeed, put does both! I simplified the insert method. 

I think if the caller really does need to know whether it exists or not then the caller can check using the new `get` method. However, I don't really think it's a problem. Would like to keep the insert method (which makes the class similar to a `set`).

Reply:
Well, actually a `std::set::insert()` will behave more like my suggestion, it will fail to insert if there is already an existing item in the set. But alright, let's keep it as it is. It's not a biggie, as long as the caller doesn't overwrite an existing `SyncConfig` in the bag by mistake (we usually opt by protect developers from coding mistakes, when possible).